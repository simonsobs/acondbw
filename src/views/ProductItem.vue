<template>
  <div class="product-item" style="position: relative;">
    <DevToolLoadingStateOverridingMenu @state="devtoolState = $event"></DevToolLoadingStateOverridingMenu>
    <div v-if="state == State.LOADING" class="mx-2 pt-5">
      <v-progress-circular indeterminate :size="26" color="grey"></v-progress-circular>
    </div>
    <div v-else-if="state == State.ERROR" class="mx-2 pt-5">
      <v-alert type="error" style="max-width: 980px;">{{ error }}</v-alert>
    </div>
    <div v-else-if="state == State.LOADED || state == State.EMPTY">
      <v-container fluid class="pa-0">
        <v-row align="start" justify="end" class="ma-0 px-0 pt-3 pb-1" style="max-width: 980px;">
          <v-tooltip bottom open-delay="800">
            <template v-slot:activator="{ on }">
              <v-btn
                text
                icon
                exact
                :to="{ name: 'ProductList', params: { productTypeName: node.type_.name } }"
                v-on="on"
              >
                <v-icon>mdi-arrow-left</v-icon>
              </v-btn>
            </template>
            <span>Back to the list</span>
          </v-tooltip>
          <v-tooltip bottom open-delay="800">
            <template v-slot:activator="{ on }">
              <v-btn icon @click="$apollo.queries.node.refetch();" v-on="on">
                <v-icon>mdi-refresh</v-icon>
              </v-btn>
            </template>
            <span>Refresh</span>
          </v-tooltip>
          <v-spacer></v-spacer>
        </v-row>
      </v-container>
      <div v-if="state == State.LOADED">
        <component
          :is="productItemCard"
          :productId="node.productId"
          :collapsible="false"
          v-on:deleted="onDeleted"
          :disableEdit="disableEdit"
          :disableDelete="disableDelete"
        ></component>
      </div>
    </div>
    <!-- <div v-else class="mx-2 pt-5">
      <v-card outlined style="max-width: 980px;">
        <v-card-text>Nothing to show here.</v-card-text>
      </v-card>
    </div>-->
  </div>
</template>

<script>
import PRODUCT_BY_TYPE_ID_AND_NAME from "@/graphql/ProductByTypeIdAndName.gql";

import ProductItemCard from "@/components/ProductItemCard";

import State from "@/utils/LoadingState.js";
import DevToolLoadingStateOverridingMenu from "@/components/DevToolLoadingStateOverridingMenu";

export default {
  name: "ProductItem",
  components: {
    ProductItemCard,
    DevToolLoadingStateOverridingMenu
  },
  props: {
    productTypeId: { required: true },
    productItemCard: { default: "ProductItemCard" },
    disableEdit: { default: false },
    disableDelete: { default: false }
  },
  data() {
    return {
      productTypeName: null,
      node: null,
      name: null,
      error: null,
      devtoolState: null,
      State: State
    };
  },
  mounted() {
    this.name = this.$route.params.name;
  },
  watch: {
    node: function() {
      if (this.node && this.node.type_) {
        this.productTypeName = this.node.type_.name;
      }
    }
  },
  computed: {
    state() {
      if (this.devtoolState) {
        return this.devtoolState;
      }

      if (this.loading) {
        return State.LOADING;
      } else if (this.error) {
        return State.ERROR;
      } else if (this.node) {
        return State.LOADED;
      } else {
        return State.NONE;
      }
    },
    loading() {
      return this.$apollo.queries.node.loading;
    }
  },
  apollo: {
    node: {
      query: PRODUCT_BY_TYPE_ID_AND_NAME,
      variables() {
        return {
          typeId: this.productTypeId,
          name: this.name
        };
      },
      skip: function() {
        return !(this.productTypeId && this.name);
      },
      update: function(data) {
        return data.product;
      },
      result(result) {
        this.error = result.error ? result.error : null;
      }
    }
  },
  methods: {
    onDeleted() {
      console.log(this.node);
      this.$router.push({
        name: "ProductList",
        params: { productTypeName: this.productTypeName }
      });
    }
  }
};
</script>
