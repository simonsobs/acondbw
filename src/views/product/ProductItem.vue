<template>
  <v-container
    fluid
    class="product-item pa-0"
    style="position: relative"
  >
    <v-row
      align="start"
      justify="end"
      class="ma-0 px-0 pt-3 pb-1"
      style="max-width: 980px"
    >
      <v-tooltip v-if="node" bottom open-delay="800">
        <template v-slot:activator="{ on }">
          <v-btn
            text
            icon
            exact
            :to="{
              name: 'ProductList',
              params: { productTypeName: node.type_.name },
            }"
            v-on="on"
          >
            <v-icon>mdi-arrow-left</v-icon>
          </v-btn>
        </template>
        <span>Back to the list</span>
      </v-tooltip>
      <v-tooltip bottom open-delay="800">
        <template v-slot:activator="{ on }">
          <v-btn :disabled="loading" icon @click="refresh()" v-on="on">
            <v-icon>mdi-refresh</v-icon>
          </v-btn>
        </template>
        <span>Refresh</span>
      </v-tooltip>
      <v-spacer></v-spacer>
    </v-row>
    <div v-if="loading" class="pa-3">
      <v-progress-circular
        indeterminate
        :size="26"
        color="grey"
      ></v-progress-circular>
    </div>
    <v-alert v-else-if="error" type="error" style="max-width: 980px">{{
      error
    }}</v-alert>
    <div v-else-if="loaded">
      <component
        :is="productItemCard"
        :productId="node.productId"
        :collapsible="false"
        v-on:deleted="onDeleted"
        :disableEdit="disableEdit"
        :disableDelete="disableDelete"
      ></component>
    </div>
    <v-card v-if="notFound" outlined style="max-width: 980px;">
      <v-card-text class="text-body-1">Not Found (404)</v-card-text>
    </v-card>
    <dev-tool-loading-state-overriding-menu
      @state="devtoolState = $event"
    ></dev-tool-loading-state-overriding-menu>
  </v-container>
</template>

<script>
import PRODUCT_BY_TYPE_ID_AND_NAME from "@/graphql/queries/ProductByTypeIdAndName.gql";

import ProductItemCard from "@/components/product/ProductItemCard.vue";

import State from "@/utils/LoadingState.js";
import DevToolLoadingStateOverridingMenu from "@/components/utils/DevToolLoadingStateOverridingMenu.vue";

export default {
  name: "ProductItem",
  components: {
    ProductItemCard,
    DevToolLoadingStateOverridingMenu,
  },
  props: {
    productTypeId: { required: true },
    productItemCard: { default: "ProductItemCard" },
    disableEdit: { default: false },
    disableDelete: { default: false },
  },
  data() {
    return {
      productTypeName: null,
      init: true,
      node: null,
      name: null,
      error: null,
      refreshing: false,
      devtoolState: null,
      State: State,
    };
  },
  mounted() {
    this.name = this.$route.params.name;
  },
  watch: {
    devtoolState: function () {
      if (this.devtoolState) {
        this.init = this.devtoolState == State.INIT;
      }
      this.error =
        this.devtoolState == State.ERROR ? "Error from Dev Tools" : null;
    },
    "$store.state.nApolloMutations": function () {
      this.refresh();
    },
    node: function () {
      if (this.node && this.node.type_) {
        this.productTypeName = this.node.type_.name;
      }
    },
  },
  computed: {
    state() {
      if (this.devtoolState) {
        return this.devtoolState;
      }

      if (this.refreshing) {
        return State.LOADING;
      }

      if (this.node) {
        return State.LOADED;
      } else if (this.$apollo.queries.node.loading) {
        return State.LOADING;
      } else if (this.error) {
        return State.ERROR;
      } else if (this.init) {
        return State.INIT;
      } else {
        return State.NONE;
      }
    },
    loading() {
      return this.state == State.LOADING;
    },
    loaded() {
      return this.state == State.LOADED;
    },
    notFound() {
      return this.state == State.NONE;
    },
  },
  apollo: {
    node: {
      query: PRODUCT_BY_TYPE_ID_AND_NAME,
      variables() {
        return {
          typeId: this.productTypeId,
          name: this.name,
        };
      },
      skip: function () {
        return !(this.productTypeId && this.name);
      },
      update: function (data) {
        return data.product;
      },
      result(result) {
        this.init = false;
        this.error = result.error ? result.error : null;
      },
    },
  },
  methods: {
    onDeleted() {
      this.$router.push({
        name: "ProductList",
        params: { productTypeName: this.productTypeName },
      });
    },
    async refresh() {
      this.refreshing = true;
      const wait = new Promise((resolve) => setTimeout(resolve, 500));
      await this.$apollo.queries.node.refetch();
      await wait; // wait until 0.5 sec passes since starting refetch
      // because the progress circular is too flickering if
      // the refetch finishes too quickly
      this.refreshing = false;
    },
  },
};
</script>
