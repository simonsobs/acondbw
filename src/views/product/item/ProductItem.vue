<template>
  <div
    class="my-5"
    style="block-size: 100%; position: relative; padding-bottom: 96px"
  >
    <div style="max-width: 960px; margin: auto">
      <top-bar :node="node" :disabled="loading" @refresh="refresh()"></top-bar>
      <div>
        <div v-if="loading">
          <v-progress-circular
            indeterminate
            :size="26"
            color="secondary"
          ></v-progress-circular>
        </div>
        <div v-else-if="error">
          <v-alert variant="tonal" type="error" :text="error"></v-alert>
        </div>
        <div v-else-if="loaded && node" class="pt-0">
          <product-item-card
            :productId="Number(node.productId)"
            :collapsible="false"
            @deleted="onDeleted"
            @nameChanged="onNameChanged($event)"
            @typeChanged="onTypeChanged($event)"
          ></product-item-card>
        </div>
        <div v-if="notFound">
          <v-card-text class="text-body-1">Not Found (404)</v-card-text>
        </div>
      </div>
    </div>
    <dev-tool-loading-state-menu top="10px" right="10px" v-model="devtoolState">
    </dev-tool-loading-state-menu>
  </div>
</template>

<script setup lang="ts">
import { ref, computed, withDefaults, onMounted } from "vue";
import { useRoute, useRouter } from "vue-router";

import { useProductByTypeIdAndNameQuery } from "@/generated/graphql";
import { useQueryState } from "@/utils/query-state";

import TopBar from "./TopBar.vue";
import ProductItemCard from "@/components/product/item-card/ProductItemCard.vue";

interface Props {
  productTypeId: number;
}

const props = defineProps<Props>();

const route = useRoute();
const router = useRouter();

const name = ref<string>("");

onMounted(() => {
  name.value = (route.params.name as string) || "";
});

const query = useProductByTypeIdAndNameQuery({
  // @ts-ignore
  variables: { typeId: props.productTypeId, name: name },
  pause: !name,
});

const node = computed(() => query.data?.value?.product);
const productTypeName = computed(() => node.value?.type_?.name);

function onDeleted() {
  if (productTypeName.value === undefined)
    throw new Error("productTypeName is undefined");
  router.push({
    name: "ProductList",
    params: { productTypeName: productTypeName.value },
  });
}

function onNameChanged(event: string) {
  if (productTypeName.value === undefined)
    throw new Error("productTypeName is undefined");
  router.push({
    name: "ProductItem",
    params: {
      productTypeName: productTypeName.value,
      name: event,
    },
  });
}

function onTypeChanged(event: string) {
  if (name.value === null) throw new Error("name is null");
  router.push({
    name: "ProductItem",
    params: {
      productTypeName: event,
      name: name.value,
    },
  });
}

const queryState = useQueryState(query, { isNull: () => node.value === null });
const { loading, loaded, notFound, error, refresh, devtoolState } = queryState;
</script>
