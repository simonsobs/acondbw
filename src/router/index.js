import Vue from "vue";
import VueRouter from "vue-router";
import VueMeta from "vue-meta";

import Frame from "@/components/layout/Frame";
import NullFrame from "@/components/layout/NullFrame";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue"

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import store from "@/store";

Vue.use(VueRouter);
Vue.use(VueMeta);

const routes = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: {
      default: OAuthRedirect,
      frame: NullFrame,
    },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: {
      default: SignIn,
      frame: NullFrame,
    },
  },
  {
    path: "/auth",
    name: "Auth",
    components: {
      default: Auth,
      frame: NullFrame,
    },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: {
      default: SignInError,
      frame: NullFrame,
    },
  },
  {
    path: "/admin/scratch",
    name: "AdminScratch",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/Scratch"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/about",
    name: "About",
    component: () =>
      import(/* webpackChunkName: "about" */ "@/views/framework/About.vue"),
  },
  {
    path: "/add-product/:productTypeName",
    components: {
      default: ProductAdd,
      frame: NullFrame,
    },
    meta: { requiresAuth: true }
  },
  {
    path: "/product/:productTypeName",
    components: {
      default: ProductTop,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    children: [
      {
        path: "",
        name: "ProductList",
        component: ProductList,
      },
      {
        path: ":name",
        name: "ProductItem",
        component: ProductItem,
      },
    ],
  },
  {
    path: "*",
    name: "NotFound",
    components: {
      default: () =>
        import(/* webpackChunkName: "error" */ "@/views/framework/NotFound.vue"),
        frame: NullFrame,
    },
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes,
});

router.beforeEach((to, from, next) => {
  const authRequired = to.matched.some((record) => record.meta.requiresAuth);
  const signedIn = !!store.state.auth.token;
  if (authRequired) {
    if (signedIn) {
      next();
    } else {
      next("/");
    }
  } else {
    next();
  }
});

export default router;
