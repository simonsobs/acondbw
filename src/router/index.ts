import Vue from "vue";
import VueRouter, { RouteConfig } from "vue-router";
import VueMeta from "vue-meta";

import { default as defaultPinia } from "@/stores";
import { useAuthStore } from "@/stores/auth";

import Frame from "@/components/layout/Frame.vue";
import NullFrame from "@/components/layout/NullFrame.vue";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";
import SearchTest from "@/views/framework/SearchTest.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue";

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import ProductType from "@/views/admin/ProductType.vue";

const About = () => import("@/views/framework/About.vue");
const NotFound = () => import("@/views/framework/NotFound.vue");
const AccessDenied = () => import("@/views/framework/AccessDenied.vue");

const FrameAdmin = () =>
  import(/* webpackChunkName: "admin" */ "@/components/admin/FrameAdmin.vue");
const Version = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Version.vue");
const Log = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Log.vue");
const Config = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Config.vue");
const Theme = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Theme.vue");
const User = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/User.vue");
const AdminAppToken = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken.vue"
  );
const AdminAppAuth = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth.vue"
  );
const AdminAppTokenError = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError.vue"
  );

Vue.use(VueRouter);
Vue.use(VueMeta);

let pinia = defaultPinia;

function setPinia(val) {
  // because not sure how to monkey patch pinia for tests
  pinia = val;
}

function setDefaultPinia() {
  setPinia(defaultPinia);
}

const routes: Array<RouteConfig> = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    // meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/search-test",
    name: "SearchTest",
    components: { default: SearchTest, frame: Frame },
    meta: { requiresAuth: true },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: { default: OAuthRedirect, frame: NullFrame },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: { default: SignIn, frame: NullFrame },
  },
  {
    path: "/auth",
    name: "Auth",
    components: { default: Auth, frame: NullFrame },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: { default: SignInError, frame: NullFrame },
  },
  {
    path: "/admin",
    name: "Admin",
    redirect: { name: "AdminVersion" },
  },
  {
    path: "/admin/product-type",
    name: "ProductType",
    components: { default: ProductType, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/versions",
    name: "AdminVersion",
    components: { default: Version, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/log",
    name: "AdminLog",
    components: { default: Log, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/config",
    name: "AdminConfig",
    components: { default: Config, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/theme",
    name: "AdminTheme",
    components: { default: Theme, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/users",
    name: "AdminUser",
    components: { default: User, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: { default: AdminAppToken, frame: FrameAdmin },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: { default: AdminAppAuth, frame: FrameAdmin },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: { default: AdminAppTokenError, frame: FrameAdmin },
  },
  {
    path: "/about",
    name: "About",
    component: About,
  },
  {
    path: "/add-product/:productTypeName",
    name: "ProductAdd",
    components: { default: ProductAdd, frame: NullFrame },
    meta: { requiresAuth: true },
  },
  {
    path: "/product/:productTypeName",
    components: { default: ProductTop, frame: Frame },
    meta: { requiresAuth: true },
    children: [
      { path: "", name: "ProductList", component: ProductList },
      { path: ":name", name: "ProductItem", component: ProductItem },
    ],
  },
  {
    path: "/access-denied",
    name: "AccessDenied",
    components: { default: AccessDenied, frame: NullFrame },
  },
  {
    path: "*",
    name: "NotFound",
    components: { default: NotFound, frame: NullFrame },
  },
];

function createRouter() {
  const router = new VueRouter({
    mode: "history",
    base: process.env.BASE_URL,
    routes,
  });

  router.beforeEach((to, from, next) => {
    const auth = useAuthStore(pinia);

    const adminRequired = to.matched.some(
      (record) => record.meta.requiresAdmin
    );
    const isAdmin = auth.isAdmin;
    if (adminRequired && !isAdmin) {
      next({ name: "AccessDenied" });
      return;
    }

    const authRequired = to.matched.some((record) => record.meta.requiresAuth);
    const isSignedIn = auth.isSignedIn as boolean;
    if (authRequired && !isSignedIn) {
      next({ name: "AccessDenied" });
      return;
    }
    next();
  });

  return router;
}

const router = createRouter();

function checkAuthForCurrentRoute() {
  const authRequired = router.currentRoute.matched.some(
    (record) => record.meta.requiresAuth
  );
  const auth = useAuthStore(pinia);
  const signedIn = auth.isSignedIn as boolean;
  if (authRequired && !signedIn) {
    router.push("/");
  }
}

export { createRouter, checkAuthForCurrentRoute, setPinia, setDefaultPinia };
