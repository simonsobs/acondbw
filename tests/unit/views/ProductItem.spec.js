import Vue from "vue";
import VueRouter from "vue-router";
import Vuetify from "vuetify";
import { mount, createLocalVue } from "@vue/test-utils";

import ProductItem from "@/views/ProductItem.vue";

import router from "@/router";

Vue.use(Vuetify);
Vue.use(VueRouter);

describe("ProductItem.vue", () => {
  let localVue;

  function createWrapper(loading = false) {
    return mount(ProductItem, {
      localVue,
      router,
      mocks: {
        $apollo: {
          queries: {
            node: {
              loading: loading,
            },
          },
        },
      },
      propsData: {
        productTypeId: 1,
        routeToProductList: { name: "MapList" },
        productItemCard: "ProductItemCard",
        nameOfRouteToProductItem: "MapItem",
      },
      stubs: {
        ProductItemCard: true,
      },
    });
  }

  beforeEach(function() {
    localVue = createLocalVue();
  });

  it("match snapshot", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      node: {
        id: "TWFwOjEwMTM=",
        productId: "1013",
        name: "lat20200201",
      },
    });
    await Vue.nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("loading state - loading", async () => {
    const loading = true;
    const wrapper = createWrapper(loading);
    await Vue.nextTick();
    expect(wrapper.find(".v-progress-circular").exists()).toBe(true);
  });

  it("loading state - error", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      error: "Error: cannot load data",
    });
    await Vue.nextTick();
    expect(wrapper.text()).toContain("Error: cannot load data");
  });

  it("loading state - none", async () => {
    const wrapper = createWrapper();
    await Vue.nextTick();
    // expect(wrapper.text()).toContain("Nothing to show here");
  });

  it("keep name when moving away", async () => {
    const wrapper = createWrapper();
    await router.push({ name: "MapItem", params: { name: "map001" } });
    await Vue.nextTick();
    expect(wrapper.vm.name).toBe("map001");

    await router.push("/about");
    await Vue.nextTick();
    expect(wrapper.vm.name).toBe("map001"); // still "map001"

    await router.push({ name: "MapItem", params: { name: "map002" } });
    await Vue.nextTick();
    expect(wrapper.vm.name).toBe("map002");
  });
});
