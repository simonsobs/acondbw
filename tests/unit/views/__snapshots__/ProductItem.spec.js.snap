// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem.vue match snapshot 1`] = `
<div class="product-item" style="position: relative;">
  <div><button type="button" class="v-btn v-btn--absolute v-btn--flat v-btn--icon v-btn--round theme--light v-size--default" style="top: -15px; right: -10px;"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-nut theme--light grey--text text--lighten-1" style="font-size: 12px;"></i></span></button>
    <div class="v-menu">
      <!---->
    </div>
  </div>
  <div>
    <div class="container pa-0 container--fluid">
      <div class="row ma-0 px-0 pt-3 pb-1 align-start justify-end" style="max-width: 980px;"><a href="/map/" class="v-btn v-btn--flat v-btn--icon v-btn--round v-btn--router v-btn--text theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-left theme--light"></i></span></a><span class="v-tooltip v-tooltip--bottom"><!----></span> <span class="v-tooltip v-tooltip--bottom"><!----></span><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-refresh theme--light"></i></span></button>
        <div class="spacer"></div>
      </div>
    </div>
    <div>
      <productitemcard-stub producttypenamesingular="map" producttypenameplural="maps" nameofroutetoproductitem="MapItem" productid="1013"></productitemcard-stub>
    </div>
  </div>
</div>
`;
